spec/push/message_apns_spec.rb

```diff
@@ -0,0 +1,135 @@
+require 'spec_helper'
+
+describe Push::MessageApns do
+  let(:notification_options) {
+    {
+      app: 'com.test.app',
+      device: 'ce8be6272e43e85516033e24b4c289220eeda4879c477160b2545e95b68b5969',
+      expiry: 1.day.to_i,
+      badge: 1,
+      attributes_for_device: { test: 1 },
+      sound: 'default',
+      alert: 'We have a new push notification for you',
+      content_available: 1,
+      priority: 10
+    }
+  }
+
+  subject { Push::MessageApns.create(notification_options) }
+
+  describe "#badge" do
+    subject { super().badge }
+    it { should == 1 }
+  end
+
+  describe "#alert" do
+    subject { super().alert }
+    it { should == "We have a new push notification for you" }
+  end
+
+  describe "#app" do
+    subject { super().app }
+    it { should == "com.test.app" }
+  end
+
+  describe "#sound" do
+    subject { super().sound }
+    it { should == "default" }
+  end
+
+  describe "#expiry" do
+    subject { super().expiry }
+    it { should == 86400 }
+  end
+
+  describe "#device" do
+    subject { super().device }
+    it { should == "ce8be6272e43e85516033e24b4c289220eeda4879c477160b2545e95b68b5969" }
+  end
+
+  describe "#content_available" do
+    subject { super().content_available }
+    it { should == 1 }
+  end
+
+  describe "#priority" do
+    subject { super().priority }
+    it { should == 10 }
+  end
+
+  describe "#attributes_for_device" do
+    subject { super().attributes_for_device }
+    it { should == { "test" => 1 } }
+  end
+
+  describe "#payload" do
+    it "should create compliant json structure" do
+      expect(subject.payload).to eq(MultiJson.dump({
+        "aps" => {
+          "alert" => "We have a new push notification for you",
+          "badge" => 1,
+          "sound" => "default",
+          "content-available" => 1
+        },
+        "test" => "1"
+      }))
+    end
+  end
+
+  describe "#to_message" do
+   it "should create a message with command 2" do
+     command, _1, _2 = subject.to_message.unpack("cNa*")
+     expect(command).to eq(2)
+   end
+
+   it "should create a message with correct frame length" do
+     _1, length, _2 = subject.to_message.unpack("cNa*")
+     expect(length).to eq(176)
+   end
+
+   def parse_items(items_stream)
+     items = []
+     until items_stream.empty?
+       item_id, item_length, items_stream = items_stream.unpack("cna*")
+       item_data, items_stream = items_stream.unpack("a#{item_length}a*")
+       items << [item_id, item_length, item_data]
+     end
+     items
+   end
+
+   it "should include five items" do
+     _1, _2, items_stream = subject.to_message.unpack("cNa*")
```

Comment:
Common pattern here could be extracted to a method or even into a custom matcher